cloudify_aws/lambda_serverless/resources/function.py

```diff
@@ -79,16 +88,44 @@ def invoke(self, params):
         '''
             Invokes an AWS Lambda Function.
         '''
-        params = params or dict()
-        params.update(dict(FunctionName=self.resource_id))
-        self.logger.debug('Invoking %s with parameters: %s'
-                          % (self.type_name, params))
-        res = self.client.invoke(**params)
+        invoke_params = dict()
+        invoke_params.update(params)
+        invoke_params.update(dict(FunctionName=self.resource_id))
+        with self._encode_payload(invoke_params.get('Payload')) as \
+                encoded_payload:
+            if encoded_payload is not None:
+                invoke_params['Payload'] = encoded_payload
+            self.logger.debug('Invoking %s with parameters: %s'
+                              % (self.type_name, invoke_params))
+            res = self.client.invoke(**invoke_params)
         if res and res.get('Payload'):
-            res['Payload'] = res['Payload'].read()
+            res['Payload'] = self._decode_payload(res['Payload'])
         self.logger.debug('Response: %s' % res)
         return res
 
+    @contextmanager
+    def _encode_payload(self, payload):
+        if isinstance(payload, str):
+            with file(payload, 'r') as payload_file:
+                yield payload_file
+        elif isinstance(payload, dict):
+            yield json.dumps(payload).encode(self.resource_encoding)
+        else:
+            yield payload
```

Comment:
@kfrukacz-gs whats the point of using the context manager here ? I mean it can be normal method

Reply:
it is used for handling the file open for the boto3 and closing when done

Reply:
I mean you do not close antyhing or do I miss anything here ? with file just close the file but what I meant the context manager does not add anything here, the method can be normal without decorator and still got the same result

Reply:
file is opened in line 109 and the stream is yielded, since it is "with" it will be closed after done